src/core/ddsi/src/ddsi_deadline.c

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright(c) 2006 to 2019 ADLINK Technology Limited and others
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Eclipse Distribution License
+ * v. 1.0 which is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
+ */
+#include <stddef.h>
+#include <stdlib.h>
+#include "dds/ddsrt/circlist.h"
+#include "dds/ddsi/ddsi_deadline.h"
+#include "dds/ddsi/q_time.h"
+#include "dds/ddsi/q_xevent.h"
+
+static void instance_deadline_missed_cb (struct xevent *xev, void *varg, nn_mtime_t tnow)
+{
+  struct deadline_adm * const deadline_adm = varg;
+  nn_mtime_t next_valid = deadline_adm->deadline_missed_cb((char *)deadline_adm - deadline_adm->list_offset, tnow);
+  resched_xevent_if_earlier (xev, next_valid);
+}
+
+/* Gets the instance from the list in deadline admin that has the earliest missed deadline and
+ * removes the instance element from the list. If no more instances with missed deadline exist
+ * in the list, the deadline (nn_mtime_t) for the first instance to 'expire' is returned. If the
+ * list is empty, NN_MTIME_NEVER is returned */
+nn_mtime_t deadline_next_missed_locked (struct deadline_adm *deadline_adm, nn_mtime_t tnow, void **instance)
+{
+  struct ddsrt_circlist_elem *list_elem = NULL;
+  struct deadline_elem *elem = NULL;
+  if (!ddsrt_circlist_isempty (&deadline_adm->list) && (list_elem = ddsrt_circlist_oldest (&deadline_adm->list)) != NULL)
+  {
+    elem = DDSRT_FROM_CIRCLIST (struct deadline_elem, e, list_elem);
+    if (elem->t_deadline.v <= tnow.v)
+    {
+      ddsrt_circlist_remove (&deadline_adm->list, &elem->e);
```

Comment:
In the normal course of events, one would reregister the instance immediately afterward, right? At which point it will be the new latest and so you're really just rotating the circular list by one position. Then again, missing deadlines would be an indication of trouble anyway, so it is good enough for now 😄 .